enhancements/microshift/microshift-pluggable-core-components.md

```diff
@@ -0,0 +1,381 @@
+
+To allow modularity of the described core components the components are split across
+multiple MicroShift subpackages, supporting at least the following use cases:
+
+| Use case 	                                        | Packages installed in the RHEL4Edge image 	| Notes 	                            |
+|--------------------------------------------------	|-----------------------------------------  	|-------	                            |
+| The whole stack of components is needed         	| `microshift`, `microshift-networking-ovnk`, `microshift-storage-topolvm` | Being explicit about the components avoids unexpected changes to customers if the default components change; i.e. they would get an error if their component of choice got deprecated. |
+| Only advanced networking is needed                | `microshift`, `microshift-networking-ovnk`  |                                    	|
+| The bare minimum set of components is necessary   | `microshift`                             |                                   	|
+| Only persistent volumes are necessary             | `microshift`, `microshift-storage-topolvm` |                                     |
+
+
+The packages for networking and storage would be different in upstrean, containing
+references to unrestricted (OKD) container images, removing the need for a pull secret.
+
+Reuse the kustomize loader so system and user components use the mechanism,
+extending its behaviour so binary plugins can be used in this content.
+
+
+`/usr/lib/microshift/components.d/004-multus/plugin`
+`/usr/lib/microshift/components.d/005-networking-ovnk/plugin`
+`/usr/lib/microshift/components.d/010-storage-topolvm/plugin`
+
+  MicroShift component manager would scan the `/usr/lib/microshift/components.d/`
+  directory executing the component plugins in alphabetical order.
+
+  `plugin` is an executable (in most cases a bash script will be the simplest
+  solution), using go binaries would be an option when complex structures
+  or connectivity back to the MicroShift API is required.
+
+  Today for the CNI/CSI components there is no wait or readiness conditions,
+  if this needed to be implemented in the future it could be implemented in
+  the `plugin`.
+
+The following changes would be introduced to packaging:
+
+* Convert `microshift` into a `microshift-core`, which does not depend
+  on `microshift-networking-ovnk`or `microshift-storage-topolvm`.
+
+* The `microshift-networking-ovnk` package would include the necessary
+  `/usr/lib/microshift/components.d/05-networking-ovnk/plugin` and collateral
+  yaml files, including CRDs.
+
+* The `microshift-storage-topolvm` package would include the necessary
+  `/usr/lib/microshift/components.d/10-storage-topolvm/plugin` and any
+   collateral yamls, including crds.
+
+* The rebase script should be updated to extract those core components
+  as and update the plugin or kustomize yamls. At some point each component
+  could implement its own rebase script.
```

Comment:
"The rebase script has to be updated..."
"At some point, each component will need to implement its own rebase script in a way that allows the component to be moved to a separate repo and potentially separate release cadence."